### PR TITLE
fix(driver-modern-bpf): fix execve USER vs MEMORY reads + cleanup + enforce upper bounds in push__bytebuf

### PR DESCRIPTION
**Pull request by incertum**
**What type of PR is this?**

> Uncomment one (or more) `/kind <>` lines:

/kind bug

/kind cleanup

> /kind design

> /kind documentation

> /kind failing-test

> /kind feature

**Any specific area of the project related to this PR?**

> Uncomment one (or more) `/area <>` lines:

> /area API-version

> /area build

> /area CI

> /area driver-kmod

> /area driver-bpf

/area driver-modern-bpf

> /area libscap-engine-bpf

> /area libscap-engine-gvisor

> /area libscap-engine-kmod

> /area libscap-engine-modern-bpf

> /area libscap-engine-nodriver

> /area libscap-engine-noop

> /area libscap-engine-source-plugin

> /area libscap-engine-savefile

> /area libscap-engine-udig

> /area libscap

> /area libpman

> /area libsinsp

> /area tests

> /area proposals

**Does this PR require a change in the driver versions?**

> /version driver-API-version-major

> /version driver-API-version-minor

> /version driver-API-version-patch

> /version driver-SCHEMA-version-major

> /version driver-SCHEMA-version-minor

> /version driver-SCHEMA-version-patch

**What this PR does / why we need it**:

Reproduce bug and verify fix:

Currently in `driver-modern-bpf` we do not get some execve data, for example:

```
sudo ./libsinsp/examples/sinsp-example --modern_bpf -f "evt.dir=< and (evt.type=execve or evt.type=execveat)" -o "* %proc.cmdline %proc.exepath %proc.name %proc.env"
```
Current master:

```
{"proc.cmdline":"git","proc.env":"","proc.exepath":"/usr/bin/git","proc.name":"git"}
```

After reverting KERNEL reads to `bpf_probe_read` or `bpf_probe_read_str` everything seems to be fixed again (this PR):

```
{"proc.cmdline":"git status","proc.env":"LONG REDACTED STRING","proc.exepath":"/usr/bin/git","proc.name":"git"}
```

> Checked in some other projects and distinguishing between KERNEL and USER reads is best practice. However, I also noticed that while `_user` bpf helpers are used, the `_kernel` helpers are not used and instead the global ones are used. Therefore, I suspect it is the actual bpf helpers causing issues here. No idea what it could be.

Edited: Actual issue was some wrong USER vs MEMORY reads attribution in execve filler.

CC folks who touched `modern_bpf` so far: @Andreagit97 @hbrueckner @Molter73 @FedeDP @jasondellaluce @leogr 

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

### CODE REVIEW

**Comment by Andreagit97** [comment]
Hey @incertum thank you for this I will take a look ASAP :)

**Comment by Andreagit97** [comment]
Just checked @incertum the fix should be simple than this, we need to change all `KERNEL` in `USER` like here https://github.com/falcosecurity/libs/blob/master/driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/execve.bpf.c#L82
We need to do this for params 2, 3 and 16 like we do in the clone.bpf.c

**Comment by hbrueckner** [comment]
Thanks @Andreagit97 

> Just checked @incertum the fix should be simple than this, we need to change all `KERNEL` in `USER` like here https://github.com/falcosecurity/libs/blob/master/driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/execve.bpf.c#L82 We need to do this for params 2, 3 and 16 like we do in the clone.bpf.c

The `arg_start` and `arg_end`, respectively for environment as well... point to user space memory. So you are right, that we need to use `USER` in those places.

**Comment by hbrueckner** [comment]
Hi @incertum 

> Hey @incertum thank you for this I will take a look ASAP :) Probably we cannot revert the changes like this because we would broken the s390x support, but yes we need to understand what's is going on here

Distinguishing between kernel and user space pointers is essential for s390x. On s390x, kernel and user space have their own address space and, hence, address can overlap. I opened #497 some time ago for the BPF driver and did a couple of reviews on modern BPF in this regard.

**Comment by incertum** [comment]
@Andreagit97 and @hbrueckner I like simpler fixes that actually make sense :smile: , squashed and adjusted last commit accordingly. Thank you!

File: driver/modern_bpf/helpers/base/push_data.h [hunk 1]
```diff
@@ -253,11 +256,20 @@ static __always_inline u16 push__charbuf(u8 *data, u64 *payload_pos, unsigned lo
  */
 static __always_inline u16 push__bytebuf(u8 *data, u64 *payload_pos, unsigned long bytebuf_pointer, u16 len_to_read, enum read_memory mem)
 {
+	u16 total_len = 0;
+	if (len_to_read > ARGS_ENV_SIZE_MAX)
```
Comment (hbrueckner) [inline driver/modern_bpf/helpers/base/push_data.h:260]: only a nit: Maybe it is worth to have some `min`/`max` macros here? (If so, maybe a follow-up PR).

**Comment by hbrueckner** [comment]
Hi @incertum 

> @Andreagit97 and @hbrueckner I like simpler fixes that actually make sense smile , squashed and adjusted last commit accordingly. Thank you!

I had also the same patch around :-)
/lgtm

File: driver/modern_bpf/helpers/base/push_data.h [hunk 2]
```diff
@@ -89,6 +89,9 @@ enum read_memory
 /* Paramater maximum size */
 #define MAX_PARAM_SIZE MAX_EVENT_SIZE - 1
 
+/* ARGS or ENV maximum size */
+#define ARGS_ENV_SIZE_MAX 4096
```
Comment (Andreagit97) [inline driver/modern_bpf/helpers/base/push_data.h:93]: not completely understood why we need this :thinking:

Comment (Andreagit97) [reply driver/modern_bpf/helpers/base/push_data.h:93]: `SAFE_ACCESS` should already protect us from the worst case

File: driver/modern_bpf/helpers/base/push_data.h (hunk 1 above)
Comment (incertum) [reply driver/modern_bpf/helpers/base/push_data.h:260]: @hbrueckner could skip reading at all if len_to_read = 0

Comment (incertum) [reply driver/modern_bpf/helpers/base/push_data.h]: @hbrueckner pushed a simple suggestion, let me know, can adjust it.

File: driver/modern_bpf/helpers/base/push_data.h (hunk 2 above)
Comment (Andreagit97) [reply driver/modern_bpf/helpers/base/push_data.h:93]: >It has been part of the old probe and it may be less for protection and rather more for you don't wanna ever log super super long unbounded strings in your logs.

That's a good point for this reason we use approaches like the `snaplen` to let users understand the max dimension they want to send, you can see an example here https://github.com/falcosecurity/libs/blob/0c280ca6847d7fbb616f152bb6cffd5b4d74452d/driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/sendto.bpf.c#L89-L98

The idea of this  `snaplen` concept is to limit at an higher level so you can chose when to limit rather then implementing it directly at low level

For what concerns the old probe we use this same concept of `snaplen` all the others checks are to please the verifier AFAIK :thinking:

File: driver/modern_bpf/helpers/base/push_data.h (hunk 1 above)
Comment (Andreagit97) [reply driver/modern_bpf/helpers/base/push_data.h]: > @hbrueckner could skip reading at all if len_to_read = 0
The code is already doing it, if `len_to_read=0` we push `0`

File: driver/modern_bpf/helpers/base/push_data.h (hunk 2 above)
Comment (incertum) [reply driver/modern_bpf/helpers/base/push_data.h:93]: Those are the two relevant max bounds from old probe

```
#define ARGS_ENV_SIZE_MAX 4096
#define MAX_PATH_LENGTH 4096
```

These values are battle proven, so wouldn't go higher before testing modern_bpf in the field.

Probably have a few options here and we can go step-wise, no need to change all now.

- Could start with uniformly enforcing bounds to existing `MAX_EVENT_SIZE`, if I can say so, the value of `2^16` appears extremely large, I think `2^12` is reasonable and enough and as mentioned battle proven. If we go this route can change this quickly and reuse the existing macro.
- If we need more granular upper bounds can introduce a new enum, similarly to the `read_memory`. For example, for the cmd args I can see that maybe `2^13` could be a good choice for an enforced upper bound. Would however perform all bound checks and selecting the corresponding upper bound here and not not in the fillers. In addition, fillers can always also enforce a lower value than the max enforced bound.

These bound checks are not unique to Falco or libs, seeing them everywhere and I would be very hesitant deploying anything to production without them.

File: driver/modern_bpf/helpers/base/push_data.h (hunk 1 above)
Comment (incertum) [reply driver/modern_bpf/helpers/base/push_data.h]: Added benefit in the modern_bpf API setup is that we exit faster here. Thought @hbrueckner raised a valid point, always be consistent and do min/max checks if you do max checks.

File: driver/modern_bpf/helpers/base/push_data.h (hunk 2 above)
Comment (incertum) [reply driver/modern_bpf/helpers/base/push_data.h:93]: Sorry typo meant the `MAX_PARAM_SIZE` you also use in `SAFE_ACCESS`

File: driver/modern_bpf/helpers/base/push_data.h (hunk 1 above)
Comment (Andreagit97) [reply driver/modern_bpf/helpers/base/push_data.h]: Uhm let's say here there is the usual trade-off:
* save one `bpf_probe_read` in case the `len_to_read` is `0`. (not frequent scenario)
* add complexity and pay for the checks when `len_to_read` is `!=0`. (frequent scenario). 

That's exactly the trade-off we have in the old probe, very complex code to save corner cases like this, to be honest, I would avoid this check at least for now, if we notice a real benefit from this we can always introduce it in a second step :thinking:

File: driver/modern_bpf/helpers/base/push_data.h (hunk 2 above)
Comment (Andreagit97) [reply driver/modern_bpf/helpers/base/push_data.h:93]: uhm I got your point maybe `MAX_PARAM_SIZE` is big enough, it's the same value we use in the old probe but it could be dangerous...

```c
/* Event maximum size */
#define MAX_EVENT_SIZE 64 * 1024

/* Paramater maximum size */
#define MAX_PARAM_SIZE MAX_EVENT_SIZE - 1
````

BTW putting a hard-coded value here doesn't seem the right solution, or better let's say it could be the right solution if we were in production and we have to rush, but this is not the case :rofl: let's take our time to think of something more dynamic or configurable

Comment (Andreagit97) [reply driver/modern_bpf/helpers/base/push_data.h:93]: I've added some dedicated checks for `execve` here https://github.com/falcosecurity/libs/pull/648/files#diff-23e1e00c90562960e7e4110fcc5c62cd89d2ea8a8df3198d7c04512084c7c378R52-R54 :)

**Review by FedeDP — COMMENTED** [review]
Hi! I'd split the "feature" `ARGS_ENV_SIZE_MAX` from the real fix, and open a PR just for that, if you agree! 
The fix is just the `driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/execve.bpf.c` diff, right?

**Comment by hbrueckner** [comment]
Hi @incertum , @FedeDP , @Andreagit97 ,

> Hi! I'd split the feature/fix `ARGS_ENV_SIZE_MAX` from the real fix, and open a PR just for that, if you agree! The fix is just the `driver/modern_bpf/programs/tail_called/events/syscall_dispatched_events/execve.bpf.c` diff, right?

trying to catch up on this and #648, my thoughts tend to have the `ARGS_ENV_SIZE_MAX` in #648 (with a co-authored-by @incertum ). The refactoring towards the collecting `execve` args/env and passing the maximum there is consistent with `MAX_PARAM_SIZE` and `MAX_UNIX_SOCKET_PATH`. And, I guess, this is for now limited to `execve` and `execveat` (the other process-creating syscalls always pass zero for now).

On [7bf2082](https://github.com/falcosecurity/libs/pull/633/commits/7bf208240cf950bada301fe1773553463eb2bbc4)](https://github.com/falcosecurity/libs/pull/633/commits/7bf208240cf950bada301fe1773553463eb2bbc4), checking `len_to_read < 1` is fine for this PR.

Wdyt?

**Comment by Andreagit97** [comment]
> trying to catch up on this and https://github.com/falcosecurity/libs/pull/648, my thoughts tend to have the ARGS_ENV_SIZE_MAX in https://github.com/falcosecurity/libs/pull/648 (with a co-authored-by @incertum ). The refactoring towards the collecting execve args/env and passing the maximum there is consistent with MAX_PARAM_SIZE and MAX_UNIX_SOCKET_PATH. And, I guess, this is for now limited to execve and execveat (the other process-creating syscalls always pass zero for now).

Sounds good :) 

> On (https://github.com/falcosecurity/libs/commit/7bf208240cf950bada301fe1773553463eb2bbc4), checking len_to_read < 1 is fine for this PR.

Yeah let's say i have only one concern about this, that's what we discovered in #648... a simple difference in how we do stuff can lead to a perf penalty not negligible, i don't think this is the case because we have just an if, but we have to say that the only gain from this `if` is that we don't call `bpf_probe_read_kernel` helper, so we save perf only in corner cases :thinking: BTW, if you think this can be valuable we can do that :)

**Comment by hbrueckner** [comment]
Hi  @incertum ,

> Hoping new commit reflects a balance @hbrueckner @Andreagit97 @FedeDP.

Commit looks fine to me.

> Small note in favor of keeping the lower bound check: no cmd args is not necessarily an edge case, especially when considering clone/fork family of syscalls. Basing this statement on some experimental observations. However, can always remove the lower bound check in the future if performance tests surface that alternatives are better.

Makes sense. Perhaps there is also a way higher in the stack to get some statistics on the such events.

/lgtm

**Review by Andreagit97 — APPROVED** [review]
Thank you for these changes :) 
/approve

**Review by FedeDP — APPROVED** [review]
/approve
